### ISSUE TITLE
Playbook hangs on [Enable and check K3s service]

### ISSUE THREAD

**Issue by nwber** [k3s-io/k3s-ansible#57]
I'm running the playbook with ` ansible-playbook site.yml -i inventory/sample/hosts.ini -k -K -vv`. 

It runs successfully up to [Enable and check K3s service] in /node/tasks/main.yml, then it hangs indefinitely. I've run this with varying levels of verbosity and debugging on. 

Running this on Raspberry Pi 4Bs, all of which have Raspberry Pi OS Lite. 

![ansible-checks-k3s-hangs](https://user-images.githubusercontent.com/68207067/87264024-e4b8c000-c48c-11ea-8e58-1dd9d998f510.png)

hosts.ini

![hosts](https://user-images.githubusercontent.com/68207067/87263491-52fc8300-c48b-11ea-809e-930f638ecfa9.PNG)

**Comment by b-m-f** [comment]
I had the same problem. 

For me the problem was the Firewall. Port 6443 was blocked on the master node.

**Comment by JohnTheNerd** [comment]
I had the same problem. For me, the master IP wasn't set properly. Anything that stops the k3s-node service from starting will cause that.

I was able to see the error by running `systemctl status k3s-node` on the node itself.

**Comment by dougbertwashere** [comment]
thanks for the info, as I  had the same problem

I assume JohnTheNerd is referring a feature of when using a hostname for the control node, the installation fails to append that info with its IPaddr to the nodes' /etc/host file

I was using a hostname in the [master] section and it failed.  Switch to an explicit  IPaddr and it installed fine.

alternate solution is to use ansible to spray out an update to the nodes' /etc/hosts file

**Comment by ShinoNuma** [comment]
I also had the same problem, and I was able to fix it by opening the port 6443 using ufw on the master node.

`sudo ufw allow 6443`

**Comment by brobare** [comment]
Thanks for the pointers in people that figured this out for their distros. Was able to resolve on Centos 7 by doing:
firewall-cmd --zone=public --add-port=6443/tcp

**Comment by bVdCreations** [comment]
I had the same problem.
I changed the inventory for the IPaddress instead of the names i set in my .ssh/config file.
that solved it!

**Comment by aireilly** [comment]
I changed `inventory/hosts.ini` to point to IP addresses instead of configured hostnames and the install worked. hostnames are listed when  I do `kubectl get nodes`:

```
NAME              STATUS   ROLES    AGE     VERSION
control-plane-0   Ready    master   3h18m   v1.17.5+k3s1
compute-node-1    Ready    <none>   4m28s   v1.17.5+k3s1
compute-node-2    Ready    <none>   4m28s   v1.17.5+k3s1
compute-node-0    Ready    <none>   4m28s   v1.17.5+k3s1
compute-node-3    Ready    <none>   4m28s   v1.17.5+k3s1
```

**Comment by jbeere** [comment]
I had this same problem, changing the hostname from "raspberrypi" to something unique seemed to make it work.

**Comment by orzen** [comment]
I had the same issue. I'm trying to deploy to a series av VPSs and it seems like the playbook is trying to use the private IP of the master node when launching the worker nodes.

**Comment by nmstoker** [comment]
Many thanks @b-m-f
I also had the firewall / port problem identified above and the following fixed it for me on Pi4s running [Raspberry Pi OS 64-bit Bullseye Lite](https://downloads.raspberrypi.org/raspios_arm64/images/) with ufw:

`sudo ufw allow proto tcp from 192.168.1.0/24 to any port 6443`

_(obviously adjust the `from` details to whatever you need for your network)_

**Comment by JohanNicander** [comment]
I had the same issue also using [Raspberry Pi OS 64-bit Bullseye Lite](https://downloads.raspberrypi.org/raspios_arm64/images/). Not until I was running with both IPs instead of hostnames and followed @nmstoker suggestion did it work. 
Many thanks!

**Comment by Lechindianer** [comment]
Just in case someone comes across this issue running Ubuntu version > 20.04: There's an issue on the k3s project regarding kernel modules: https://github.com/k3s-io/k3s/issues/4234#issuecomment-947954002

Installing the necessary kernel modules helped me, the playbook ran successfully without hanging

**Comment by Zedifuu** [comment]
@Lechindianer You absolute champion, that was my issue! TY so much!

**Comment by ozbillwang** [comment]
the task `Enable and check K3s service` is to restart / enable `k3s-node` service in nodes

```
- name: Enable and check K3s service
  systemd:
    name: k3s-node
    daemon_reload: yes
    state: restarted
    enabled: yes
```

the problem is related to the setting here:

in file `roles/k3s/node/templates/k3s.service.j2`

```
ExecStart=/usr/local/bin/k3s agent --server https://{{ master_ip }}:6443 --token {{ hostvars[groups['master'][0]]['token'] }} {{ extra_agent_args | default("") }}
```

But the variable `master_ip` is not always master's IP, it is from `inventory/my-cluster/hosts.ini`.  If you put IP in `hosts.ini`, then you get IP address, if you put hostname in it, you get hostname

```
inventory/my-cluster/group_vars/all.yml:master_ip: "{{ hostvars[groups['master'][0]]['ansible_host'] | default(groups['master'][0]) }}"
```

But in my case, I set the hostname, not IP in the `hosts.ini`, as poster, like this:

```
[master]
kubernetes-master
```

So when check in node, I got 

```
ExecStart=/usr/local/bin/k3s agent --server https://kubernetes-master:6443 --token xxx
```

But there is no /etc/hosts to resolve its IP address, so the playbook is wait to join to the master, never finish

**Comment by OladapoAjala** [comment]
Thanks @ozbillwang for the detailed info, I however fixed this by slightly modifying your suggestion.

In the `hosts.ini` file I added a new variable

```ini
[master]
HOST_NAME ansible_host_ip=192.168.xxx.xxx
```

then in the /group_vars/all.yml file, I changed how the variable `master_ip` is being resolved by replacing `ansible_host` with `ansible_host_ip` which is already set above.

**Comment by filipemeneses** [comment]
Editing the nodes password at `/etc/rancher/node/password` to match the master node worked for me

**Comment by dereknola** [comment]
Closing this as discussion seems to have ended. It is recommended that 
1) All nodes have static IPs OR an external load balancer is configured with a fixed registration address. See https://docs.k3s.io/datastore/ha#4-optional-configure-a-fixed-registration-address and https://docs.k3s.io/datastore/cluster-loadbalancer
2) Firewalls be disabled. I have another issue open to track attempting to open the firewall ports. https://github.com/k3s-io/k3s-ansible/issues/234
